File: Examples/VisualBasic/ModifyingAndConvertingImages/ReadAndWriteXMPDataToImages.vb
Imports System.IO
Imports Aspose.Imaging.FileFormats.Tiff
Imports Aspose.Imaging.FileFormats.Tiff.Enums
Imports Aspose.Imaging.ImageOptions
Imports Aspose.Imaging.Xmp
Imports Aspose.Imaging.Xmp.Schemas.DublinCore
Imports Aspose.Imaging.Xmp.Schemas.Photoshop

'
'This project uses Automatic Package Restore feature of NuGet to resolve Aspose.Imaging for .NET API reference 
'when the project is build. Please check https://Docs.nuget.org/consume/nuget-faq for more information. 
'If you do not wish to use NuGet, you can manually download Aspose.Imaging for .NET API from http://www.aspose.com/downloads, 
'Install it and then add its reference to this project. For any issues, questions or suggestions 
'Please feel free to contact us using http://www.aspose.com/community/forums/default.aspx
'

Namespace Aspose.Imaging.Examples.VisualBasic.ModifyingAndConvertingImages
    Class ReadAndWriteXMPDataToImages
        Public Shared Sub Run()
            ' ExStart:ReadAndWriteXMPDataToImages
            ' The path to the documents directory.
            Dim dataDir As String = RunExamples.GetDataDir_ModifyingAndConvertingImages()

            ' Specify the size of image by defining a Rectangle 
            Dim rect As New Rectangle(0, 0, 100, 200)
            Dim tiffOptions As New TiffOptions(TiffExpectedFormat.TiffJpegRgb)
            tiffOptions.Photometric = TiffPhotometrics.MinIsBlack
            tiffOptions.BitsPerSample = New UShort() {8}

            ' Create the brand new image just for sample purposes
            Using image__1 = New TiffImage(New TiffFrame(tiffOptions, rect.Width, rect.Height))
                ' Create an instance of XMP-Header and Xmp-TrailerPi 
                Dim xmpHeader As New XmpHeaderPi(Guid.NewGuid().ToString())
                Dim xmpTrailer As New XmpTrailerPi(True)

                ' Create an instance of XMPmeta class to set different attributes
                Dim xmpMeta As New XmpMeta()
                xmpMeta.AddAttribute("Author", "Mr Smith")
                xmpMeta.AddAttribute("Description", "The fake metadata value")

                ' Create an instance of XmpPacketWrapper that contains all metadata
                Dim xmpData As New XmpPacketWrapper(xmpHeader, xmpTrailer, xmpMeta)

                ' Create an instacne of Photoshop package and set photoshop attributes
                Dim photoshopPackage As New PhotoshopPackage()
                photoshopPackage.SetCity("London")
                photoshopPackage.SetCountry("England")
                photoshopPackage.SetColorMode(ColorMode.Rgb)
                photoshopPackage.SetCreatedDate(DateTime.UtcNow)

                ' Add photoshop package into XMP metadata
                xmpData.AddPackage(photoshopPackage)

                ' Create an instacne of DublinCore package and set dublinCore attributes
                Dim dublinCorePackage As New DublinCorePackage()
                dublinCorePackage.SetAuthor("Charles Bukowski")
                dublinCorePackage.SetTitle("Confessions of a Man Insane Enough to Live With the Beasts")
                dublinCorePackage.AddValue("dc:movie", "Barfly")

                ' Add dublinCore Package into XMP metadata
                xmpData.AddPackage(dublinCorePackage)

                Using ms = New MemoryStream()
                    ' Update XMP metadata into image and Save image on the disk or in memory stream
                    image__1.XmpData = xmpData
                    image__1.Save(ms)
                    ms.Seek(0, System.IO.SeekOrigin.Begin)
                    ' Load the image from moemory stream or from disk to read/get the metadata
                    Using img = DirectCast(Image.Load(ms), TiffImage)
                        ' Getting the XMP metadata
                        Dim imgXmpData As XmpPacketWrapper = img.XmpData
                        ' Use package data ...
                        For Each package As XmpPackage In imgXmpData.Packages
                        Next
                    End Using
                End Using
            End Using
        End Sub
        ' ExEnd:ReadAndWriteXMPDataToImages
    End Class
End Namespace

File: Examples/CSharp/ModifyingAndConvertingImages/ReadAndWriteXMPDataToImages.cs
// GIST-ID: 43eebf03ec07028adf45b293105b7cd1
using System;
using System.IO;
using Aspose.Imaging;
using Aspose.Imaging.FileFormats.Tiff;
using Aspose.Imaging.FileFormats.Tiff.Enums;
using Aspose.Imaging.ImageOptions;
using Aspose.Imaging.Xmp;
using Aspose.Imaging.Xmp.Schemas.DublinCore;
using Aspose.Imaging.Xmp.Schemas.Photoshop;

/*
This project uses the Automatic Package Restore feature of NuGet to resolve the Aspose.Imaging for .NET API reference
when the project is built. Please check https://learn.microsoft.com/en-us/nuget/resources/nuget-faq for more information.
If you do not wish to use NuGet, you can manually download Aspose.Imaging for .NET API from https://releases.aspose.com/,
install it, and then add its reference to this project. For any issues, questions, or suggestions,
please feel free to contact us using https://forum.aspose.com/
*/

namespace Aspose.Imaging.Examples.CSharp.ModifyingAndConvertingImages
{
    class ReadAndWriteXMPDataToImages
    {
        public static void Run()
        {
            Console.WriteLine("Running example ReadAndWriteXMPDataToImages");
            // The path to the documents directory.
            string dataDir = RunExamples.GetDataDir_ModifyingAndConvertingImages();

            // Specify the size of the image by defining a Rectangle.
            Rectangle rect = new Rectangle(0, 0, 100, 200);
            TiffOptions tiffOptions = new TiffOptions(TiffExpectedFormat.TiffJpegRgb);
            tiffOptions.Photometric = TiffPhotometrics.MinIsBlack;
            tiffOptions.BitsPerSample = new ushort[] { 8 };

            // Create a brand‑new image just for sample purposes.
            using (var image = new TiffImage(new TiffFrame(tiffOptions, rect.Width, rect.Height)))
            {
                // Create an instance of XMP header.
                XmpHeaderPi xmpHeader = new XmpHeaderPi(Guid.NewGuid().ToString());

                // Create an instance of XmpTrailerPi and an XmpMeta object to set attributes.
                XmpTrailerPi xmpTrailer = new XmpTrailerPi(true);
                XmpMeta xmpMeta = new XmpMeta();
                xmpMeta.AddAttribute("Author", "Mr Smith");
                xmpMeta.AddAttribute("Description", "The fake metadata value");

                // Create an instance of XmpPacketWrapper that contains all metadata.
                XmpPacketWrapper xmpData = new XmpPacketWrapper(xmpHeader, xmpTrailer, xmpMeta);

                // Create an instance of Photoshop package and set Photoshop attributes.
                PhotoshopPackage photoshopPackage = new PhotoshopPackage();
                photoshopPackage.SetCity("London");
                photoshopPackage.SetCountry("England");
                photoshopPackage.SetColorMode(ColorMode.Rgb);
                photoshopPackage.SetCreatedDate(DateTime.UtcNow);

                // Add Photoshop package into XMP metadata.
                xmpData.AddPackage(photoshopPackage);

                // Create an instance of DublinCore package and set DublinCore attributes.
                DublinCorePackage dublinCorePackage = new DublinCorePackage();
                dublinCorePackage.SetAuthor("Charles Bukowski");
                dublinCorePackage.SetTitle("Confessions of a Man Insane Enough to Live With the Beasts");
                dublinCorePackage.AddValue("dc:movie", "Barfly");

                // Add DublinCore package into XMP metadata.
                xmpData.AddPackage(dublinCorePackage);

                using (var ms = new MemoryStream())
                {
                    // Update XMP metadata into the image and save the image on disk or in the memory stream.
                    image.XmpData = xmpData;
                    image.Save(ms);
                    ms.Seek(0, System.IO.SeekOrigin.Begin);

                    // Load the image from the memory stream or from disk to read/get the metadata.
                    using (var img = (TiffImage)Image.Load(ms))
                    {
                        // Getting the XMP metadata.
                        XmpPacketWrapper imgXmpData = img.XmpData;
                        foreach (XmpPackage package in imgXmpData.Packages)
                        {
                            // Use package data ...
                        }
                    }
                }
            }

            Console.WriteLine("Finished example ReadAndWriteXMPDataToImages");
        }
    }
}